src/qibo/backends/numpy.py

```diff
@@ -769,12 +769,32 @@ def calculate_matrix_exp(self, a, matrix, eigenvectors=None, eigenvalues=None):
         ud = self.np.transpose(np.conj(eigenvectors))
         return self.np.matmul(eigenvectors, self.np.matmul(expd, ud))
 
-    def calculate_matrix_power(self, matrix, power: Union[float, int]):
+    def calculate_matrix_power(
+        self,
+        matrix,
+        power: Union[float, int],
+        precision_singularity: float = 1e-14,
+    ):
         if not isinstance(power, (float, int)):
             raise_error(
                 TypeError,
                 f"``power`` must be either float or int, but it is type {type(power)}.",
             )
+
+        if power < 0.0:
+            # negative powers of singular matrices via SVD
+            determinant = self.np.linalg.det(matrix)
+            if abs(determinant) < precision_singularity:
+                U, S, Vh = self.np.linalg.svd(matrix)
+                # cast needed because of different dtypes in `torch`
+                S = self.cast(S)
+                S_inv = self.np.where(
+                    self.np.abs(S) < precision_singularity, 0.0, S**power
+                )
+                return (
+                    self.np.linalg.inv(Vh) @ self.np.diag(S_inv) @ self.np.linalg.inv(U)
+                )
```

Comment:
I didn't check all the details, so maybe I am wrong, but it seems that maybe we could reduce code repetition by moving this to a function that accepts `self.np` as an argument and then reuse for tensorflow (passing `self.tf`) and maybe even for cupy in qibojit.

I mean something like
```py
def calculate_matrix_negative_power(mod, matrix, ...):
    determinant = mod.linalg.det(matrix)
    ...
```
It can be defined even outside the backend, but then it will need to be imported in the various places, otherwise as `@staticmethod` in the backend.